File: push-be/internal/domain/usecase/operation.go
package usecase

import (
	"context"
	"errors"
	"time"

	"github.com/google/uuid"

	"push-be/internal/domain"
)

type OperationService struct {
	ops      domain.OperationStore
	uow      domain.UnitOfWork
	evidence *EvidenceService
}

func NewOperationService(ops domain.OperationStore, uow domain.UnitOfWork, evidence *EvidenceService) *OperationService {
	return &OperationService{ops: ops, uow: uow, evidence: evidence}
}

func (s *OperationService) Get(ctx context.Context, userID, id uuid.UUID) (*domain.Operation, error) {
	op, err := s.ops.Get(ctx, userID, id)
	if err != nil {
		if errors.Is(err, domain.ErrNotFound) {
			return nil, domain.NotFound()
		}
		return nil, domain.Internal()
	}
	return op, nil
}

func (s *OperationService) SubmitInput(ctx context.Context, userID, id uuid.UUID, fields map[string]string, sourceID *uuid.UUID) (*domain.Operation, error) {
	var out *domain.Operation
	err := s.uow.Do(ctx, func(ctx context.Context) error {
		op, err := s.ops.Get(ctx, userID, id)
		if err != nil {
			if errors.Is(err, domain.ErrNotFound) {
				return domain.NotFound()
			}
			return err
		}
		if domain.TerminalOperationStatus(op.Status) {
			return domain.InvalidTransition("operation already finished")
		}
		if op.Status != domain.OpNeedsInput || op.InputRequest == nil {
			return domain.InvalidTransition("operation does not await input")
		}
		switch op.Type {
		case domain.OpEvidenceImport:
			if err := s.evidence.CompleteImportInput(ctx, op, fields); err != nil {
				return err
			}
		default:
			return domain.InvalidTransition("operation does not accept input")
		}
		op.InputRequest = nil
		op.UpdatedAt = time.Now().UTC()
		if op.Status == domain.OpNeedsInput {
			op.Status = domain.OpQueued
		}
		if err := s.ops.Update(ctx, op); err != nil {
			return err
		}
		out = op
		return nil
	})
	if err != nil {
		return nil, err
	}
	return out, nil
}

func (s *OperationService) Cancel(ctx context.Context, userID, id uuid.UUID) (*domain.Operation, error) {
	var out *domain.Operation
	err := s.uow.Do(ctx, func(ctx context.Context) error {
		op, err := s.ops.Get(ctx, userID, id)
		if err != nil {
			if errors.Is(err, domain.ErrNotFound) {
				return domain.NotFound()
			}
			return err
		}
		if domain.TerminalOperationStatus(op.Status) {
			out = op
			return nil
		}
		op.Status = domain.OpCancelled
		op.UpdatedAt = time.Now().UTC()
		if err := s.ops.Update(ctx, op); err != nil {
			return err
		}
		out = op
		return nil
	})
	if err != nil {
		return nil, err
	}
	return out, nil
}

File: push-be/app/domain/services/operation.py
from __future__ import annotations
from datetime import datetime, timezone
from typing import Optional
from uuid import UUID

from app.db import DB, NotFoundError
from app.domain import entities as ent
from app.domain.errors import internal, invalid_transition, not_found
from app.infrastructure.store_operations import OperationStore


def _now() -> datetime:
    return datetime.now(timezone.utc)


class OperationService:
    def __init__(self, db: DB, evidence_svc):
        self.db = db
        self.ops = OperationStore(db)
        self.evidence = evidence_svc

    async def get(self, user_id: UUID, id_: UUID) -> ent.Operation:
        try:
            return await self.ops.get(user_id, id_)
        except NotFoundError:
            raise not_found()
        except Exception:
            raise internal()

    async def submit_input(self, user_id: UUID, id_: UUID, fields: dict,
                           source_id: Optional[UUID]) -> ent.Operation:
        out = None

        async def work():
            nonlocal out
            try:
                op = await self.ops.get(user_id, id_)
            except NotFoundError:
                raise not_found()
            if ent.terminal_operation_status(op.status):
                raise invalid_transition("operation already finished")
            if op.status != ent.OP_NEEDS_INPUT or op.input_request is None:
                raise invalid_transition("operation does not await input")
            if op.type == ent.OP_EVIDENCE_IMPORT:
                await self.evidence.complete_import_input(op, fields)
            else:
                raise invalid_transition("operation does not accept input")
            op.input_request = None
            op.updated_at = _now()
            if op.status == ent.OP_NEEDS_INPUT:
                op.status = ent.OP_QUEUED
            await self.ops.update(op)
            out = op

        await self.db.do(work)
        return out

    async def cancel(self, user_id: UUID, id_: UUID) -> ent.Operation:
        out = None

        async def work():
            nonlocal out
            try:
                op = await self.ops.get(user_id, id_)
            except NotFoundError:
                raise not_found()
            if ent.terminal_operation_status(op.status):
                out = op
                return
            op.status = ent.OP_CANCELLED
            op.updated_at = _now()
            await self.ops.update(op)
            out = op

        await self.db.do(work)
        return out
